src/parseTree.js

```diff
@@ -281,37 +270,121 @@
     props: {
       traceNode: {type: Object, required: true},
       measureInputText: {type: Function},
-      classes: {type: Object, default: Object},
       isInVBox: {type: Boolean},
-      vbox: {type: Boolean}
+
+      // from parent element
+      context: {type: Object},
+
+      // from currentLR
+      currentLR: {type: Object, default: Object},
+      eventHandlers: {type: Object}
     },
     computed: {
+      labeled: function() {
+        return shouldNodeBeLabeled(this.traceNode);
+      },
+      vbox: function() {
+        return hasVisibleChoice(this.traceNode) ||
+               hasVisibleLeftRecursion(this.traceNode) ||
+               (isAlt(this.traceNode.expr) && this.context && this.context.vbox);
+      },
+      isWhitespace: function() {
+        return this.traceNode.ruleName === 'spaces';
+      },
+      isLeaf: function() {
+        var leaf = isLeaf(this.traceNode);
+        if (this.traceNode.isMemoized) {
+          var memoKey = this.traceNode.expr.toMemoKey();
+          var stack = this.currentLR[memoKey];
+          if (stack && stack[stack.length - 1] === this.traceNode.pos) {
+            leaf = true;
+          }
+        }
+        return leaf;
+      },
       classObj: function() {
         var obj = {
-          disclosure: this.classes.labeled && this.isInVBox
+          disclosure: this.labeled && this.isInVBox
         };
         var ctorName = this.traceNode.ctorName;
         if (ctorName) {
           obj[ctorName.toLowerCase()] = true;
         }
-        Object.assign(obj, this.classes);
+
+        obj.collapsed = this.labeled &&
```

Comment:
Oh, yeah -- sorry, you're right. Then let's just keep this as is, and if we want to refactor it later, we can.